community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/SchemaCache.java

```diff
@@ -55,16 +52,16 @@
  */
 public class SchemaCache
 {
-    private final Map<Long, IndexRule> indexRuleById = new HashMap<>();
-    private final Map<Long, ConstraintRule> constraintRuleById = new HashMap<>();
-    private final Set<ConstraintDescriptor> constraints = new HashSet<>();
+    private final Map<Long,IndexRule> indexRuleById = new ConcurrentHashMap<>();
+    private final Map<Long,ConstraintRule> constraintRuleById = new ConcurrentHashMap<>();
+    private final Map<Long,ConstraintDescriptor> constraints = new ConcurrentHashMap<>();
 
-    private final Map<SchemaDescriptor,IndexDescriptor> indexDescriptors = new HashMap<>();
-    private final PrimitiveIntObjectMap<Set<IndexDescriptor>> indexDescriptorsByLabel = Primitive.intObjectMap();
+    private final Map<SchemaDescriptor,IndexDescriptor> indexDescriptors = new ConcurrentHashMap<>();
+    private final Map<Integer,Set<IndexDescriptor>> indexDescriptorsByLabel = new ConcurrentHashMap<>();
     private final ConstraintSemantics constraintSemantics;
 
     private final Map<Class<?>,Object> dependantState = new ConcurrentHashMap<>();
-    private final PrimitiveIntObjectMap<List<IndexDescriptor>> indexByProperty = Primitive.intObjectMap();
+    private final Map<Integer,List<IndexDescriptor>> indexByProperty = new ConcurrentHashMap<>();
```

Comment:
I think we should have a card for looking into an implementation of this based on snapshotting and atomic multi-value updates.

Reply:
I think so too.